File: _inc/laravel/app/Http/Controllers/Activity/LeadStageController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{MiddlewaresConstants, PermissionsConstants};
use App\Models\{LeadStage, Pipeline};
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Http\{JsonResponse, RedirectResponse, Request};
use Illuminate\Support\Facades\{Auth, Log, Route, Validator, View as ViewFacade};
use Illuminate\View\View;
use Symfony\Component\HttpFoundation\Response;

use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
use App\Traits\HasCrudConstants;
use App\Traits\DefinesResourceActions;
class LeadStageController extends Controller
{
	use DefinesResourceActions;

    use HasCrudConstants;

  public function __construct()
  {
    $this->middleware([MiddlewaresConstants::AUTH, MiddlewaresConstants::XSS]);
  }

  public function index(Request $req): RedirectResponse|JsonResponse|View
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = 'lead_stages.index';
    return $this->measureProfile($action, function () use ($req, $action, $class, $viewPath) {
      Log::info("[{$class}::{$action}] start", ['owner_id' => $req->user()?->ownerId()]);
      if ($r = self::_deny($req, PermissionsConstants::MNG_LD_ST)) return $r;
      try {
        $ownerId = $req->user()->ownerId();
        $fetchStart = microtime(true);
        $leadStages = LeadStage::with('pipeline')->where('lead_stages.created_by', $ownerId)->whereHas('pipeline', fn($q) => $q->where('created_by', $ownerId))->orderBy('pipeline_id')->orderBy('order')->get();
        $this->logExecutionTime($fetchStart, $action, 'fetchLeadStages');
        $groupStart = microtime(true);
        $pipelines = $leadStages->groupBy('pipeline_id')->map(fn($c) => ['name' => $c->first()->pipeline->name, 'leadStages' => $c])->all();
        $this->logExecutionTime($groupStart, $action, 'groupLeadStages');
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        Log::info("[{$class}::{$action}] complete", ['lead_stage_count' => $leadStages->count(), 'pipeline_count' => count($pipelines)]);
        return view($viewPath, compact('pipelines'));
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'owner_id' => $req->user()?->ownerId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public function create(Request $req): RedirectResponse|JsonResponse|View
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = 'lead_stages.create';
    return $this->measureProfile($action, function () use ($req, $action, $class, $viewPath) {
      Log::info("[{$class}::{$action}] start", ['owner_id' => $req->user()?->ownerId()]);
      if ($r = self::_deny($req, 'create lead stage')) return $r;
      try {
        $ownerId = $req->user()->ownerId();
        $fetchStart = microtime(true);
        $pipelines = Pipeline::where('created_by', $ownerId)->pluck('name', 'id');
        $this->logExecutionTime($fetchStart, $action, 'fetchPipelines');
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        Log::info("[{$class}::{$action}] complete", ['pipeline_count' => $pipelines->count()]);
        return view($viewPath, compact('pipelines'));
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'owner_id' => $req->user()?->ownerId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public function store(Request $req): RedirectResponse|JsonResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    return $this->measureProfile($action, function () use ($req, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['owner_id' => $req->user()?->ownerId(), 'input_keys' => array_keys($req->all())]);
      if ($r = self::_deny($req, 'create lead stage')) return $r;
      if ($v = self::_validate($req->all(), ['name' => 'required|max:20', 'pipeline_id' => 'required'])) return $v;
      try {
        $createStart = microtime(true);
        LeadStage::create(['name' => $req->name, 'pipeline_id' => $req->pipeline_id, 'created_by' => $req->user()->ownerId()]);
        $this->logExecutionTime($createStart, $action, 'createLeadStage');
        Log::info("[{$class}::{$action}] success", ['name' => $req->name, 'pipeline_id' => $req->pipeline_id]);
        return redirect()->route('lead_stages.index')->with('success', 'Lead Stage successfully created!');
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'owner_id' => $req->user()?->ownerId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public function show(Request $req, LeadStage $leadStage): RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    return $this->measureProfile($action, function () use ($req, $leadStage, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['lead_stage_id' => $leadStage->getKey()]);
      try {
        $redirStart = microtime(true);
        $resp = redirect()->route('lead_stages.index');
        $this->logExecutionTime($redirStart, $action, 'redirectToIndex');
        Log::info("[{$class}::{$action}] complete");
        return $resp;
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['lead_stage_id' => $leadStage->getKey(), 'route' => Route::getCurrentRoute()?->getName(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'lead_stage_id' => $leadStage->getKey()]);
  }

  public function edit(Request $req, LeadStage $leadStage): RedirectResponse|JsonResponse|View
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = 'lead_stages.edit';
    return $this->measureProfile($action, function () use ($req, $leadStage, $action, $class, $viewPath) {
      Log::info("[{$class}::{$action}] start", ['lead_stage_id' => $leadStage->getKey(), 'owner_id' => $req->user()?->ownerId()]);
      if ($r = self::_deny($req, 'edit lead stage')) return $r;
      if ($leadStage->created_by !== $req->user()->ownerId()) return defaultPermissionDenial($req, new AuthorizationException('lead owner'), $class . '::' . $action);
      try {
        $ownerId = $req->user()->ownerId();
        $fetchStart = microtime(true);
        $pipelines = Pipeline::where('created_by', $ownerId)->pluck('name', 'id');
        $this->logExecutionTime($fetchStart, $action, 'fetchPipelines');
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        Log::info("[{$class}::{$action}] complete", ['pipeline_count' => $pipelines->count()]);
        return view($viewPath, compact('leadStage', 'pipelines'));
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'lead_stage_id' => $leadStage->getKey(), 'owner_id' => $req->user()?->ownerId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'lead_stage_id' => $leadStage->getKey()]);
  }

  public function update(Request $req, LeadStage $leadStage): RedirectResponse|JsonResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    return $this->measureProfile($action, function () use ($req, $leadStage, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['lead_stage_id' => $leadStage->getKey(), 'owner_id' => $req->user()?->ownerId(), 'input_keys' => array_keys($req->all())]);
      if ($r = self::_deny($req, 'edit lead stage')) return $r;
      if ($leadStage->created_by !== $req->user()->ownerId()) return defaultPermissionDenial($req, new AuthorizationException('lead owner'), $class . '::' . $action);
      if ($v = self::_validate($req->all(), ['name' => 'required|max:20', 'pipeline_id' => 'required'])) return $v;
      try {
        $updateStart = microtime(true);
        $leadStage->update(['name' => $req->name, 'pipeline_id' => $req->pipeline_id]);
        $this->logExecutionTime($updateStart, $action, 'updateLeadStage');
        Log::info("[{$class}::{$action}] success", ['lead_stage_id' => $leadStage->getKey(), 'name' => $req->name, 'pipeline_id' => $req->pipeline_id]);
        return redirect()->route('lead_stages.index')->with('success', 'Lead Stage successfully updated!');
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'lead_stage_id' => $leadStage->getKey(), 'owner_id' => $req->user()?->ownerId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'lead_stage_id' => $leadStage->getKey()]);
  }

  public function destroy(Request $req, LeadStage $leadStage): RedirectResponse|JsonResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    return $this->measureProfile($action, function () use ($req, $leadStage, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['lead_stage_id' => $leadStage->getKey(), 'owner_id' => $req->user()?->ownerId()]);
      if ($r = self::_deny($req, 'delete lead stage')) return $r;
      try {
        $deleteStart = microtime(true);
        $leadStage->delete();
        $this->logExecutionTime($deleteStart, $action, 'deleteLeadStage');
        Log::info("[{$class}::{$action}] success", ['lead_stage_id' => $leadStage->getKey()]);
        return redirect()->route('lead_stages.index')->with('success', 'Lead Stage successfully deleted!');
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'lead_stage_id' => $leadStage->getKey(), 'owner_id' => $req->user()?->ownerId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'lead_stage_id' => $leadStage->getKey()]);
  }

  public function order(Request $req): RedirectResponse|JsonResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    return $this->measureProfile($action, function () use ($req, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['owner_id' => $req->user()?->ownerId(), 'payload_count' => is_array($req->input('order')) ? count($req->input('order')) : 0]);
      if ($r = self::_deny($req, 'edit lead stage')) return $r;
      $payload = $req->input('order', []);
      try {
        $updStart = microtime(true);
        foreach ($payload as $idx => $id) LeadStage::where('id', $id)->update(['order' => $idx]);
        $this->logExecutionTime($updStart, $action, 'bulkUpdateOrder');
        Log::info("[{$class}::{$action}] success", ['updated' => count($payload)]);
        return response()->json(['status' => 'ok'], Response::HTTP_OK);
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'owner_id' => $req->user()?->ownerId(), 'payload_count' => is_array($payload) ? count($payload) : 0, 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'payload_count' => is_array($req->input('order')) ? count($req->input('order')) : 0]);
  }

  private static function _deny(
    Request $req,
    string  $perm
  ): RedirectResponse|JsonResponse|null {
    return $req->user()->can($perm)
      ? null
      : defaultPermissionDenial(
        $req,
        new \Illuminate\Auth\Access\AuthorizationException($perm),
        __CLASS__ . '::' . debug_backtrace(
          DEBUG_BACKTRACE_IGNORE_ARGS,
          2
        )[1]['function']
      );
  }

  private static function _validate(
    array $data,
    array $rules
  ): ?RedirectResponse {
    $v = Validator::make($data, $rules);
    if ($v->fails()) {
      return redirect()
        ->back()
        ->with('error', $v->getMessageBag()->first());
    }
    return null;
  }
}

// ! ALERT order() trusts client‑supplied indexes; consider validating that each ID belongs to the authenticated owner and wrapping updates in a DB transaction with row‑level locking to prevent race conditions.

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/activity/lead_stage_controller.py
import logging
import inspect
from django.contrib import messages
from django.core.exceptions import PermissionDenied
from django.db import transaction
from django.http import HttpRequest, HttpResponse, JsonResponse
from django.shortcuts import get_object_or_404, redirect, render
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from ....Models.activity.lead_stage import LeadStage
from ....Models.configs.pipeline import Pipeline
from .._traits.controller import Controller

logger = logging.getLogger(__name__)

class LeadStageController(Controller):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.middleware(['auth','XSS'])

    @classmethod
    def index(cls, request: HttpRequest) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('manage lead stage'):
                raise PermissionDenied('User lacks permission: manage lead stage')
            qs = LeadStage.objects.select_related('pipeline').filter(
                pipeline__created_by=request.user.owner_id(),
                created_by=request.user.owner_id()
            ).order_by('pipeline_id','order')
            pipelines = {}
            for ls in qs:
                pid = ls.pipeline_id
                if pid not in pipelines:
                    pipelines[pid] = {'name': ls.pipeline.name, 'lead_stages': []}
                pipelines[pid]['lead_stages'].append(ls)
            return render(request, 'lead_stages/index.html', {'pipelines': pipelines})
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=REF, logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=REF, logger=logger)

    @classmethod
    def create(cls, request: HttpRequest) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('create lead stage'):
                raise PermissionDenied('User lacks permission: create lead stage')
            pipelines = Pipeline.objects.filter(
                created_by=request.user.owner_id()
            ).values_list('name','id')
            return render(request, 'lead_stages/create.html', {'pipelines': pipelines})
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=REF, logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=REF, logger=logger)

    @classmethod
    @transaction.atomic
    def store(cls, request: HttpRequest) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('create lead stage'):
                raise PermissionDenied('User lacks permission: create lead stage')
            data = request.POST
            name = data.get('name')
            pid = data.get('pipeline_id')
            if not name or len(name) > 20 or not pid:
                messages.error(request, 'Name and Pipeline are required.')
                return redirect('lead_stages_index')
            ls = LeadStage(name=name, pipeline_id=pid, created_by=request.user.owner_id())
            ls.save()
            messages.success(request, 'Lead Stage successfully created!')
            return redirect('lead_stages_index')
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=REF, logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=REF, logger=logger)

    @classmethod
    def edit(cls, request: HttpRequest, lead_stage_id: int) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('edit lead stage'):
                raise PermissionDenied('User lacks permission: edit lead stage')
            ls = get_object_or_404(LeadStage, pk=lead_stage_id)
            if ls.created_by != request.user.owner_id():
                return default_permission_denial(request, err=PermissionDenied(), ref=REF, logger=logger, json={})
            pipelines = Pipeline.objects.filter(
                created_by=request.user.owner_id()
            ).values_list('name','id')
            return render(request, 'lead_stages/edit.html', {
                'leadStage': ls,
                'pipelines': pipelines
            })
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=REF, logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=REF, logger=logger)

    @classmethod
    def update(cls, request: HttpRequest, lead_stage_id: int) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('edit lead stage'):
                raise PermissionDenied('User lacks permission: edit lead stage')
            ls = get_object_or_404(LeadStage, pk=lead_stage_id)
            if ls.created_by != request.user.owner_id():
                raise PermissionDenied('User lacks permission: edit lead stage')
            data = request.POST
            name = data.get('name')
            pid = data.get('pipeline_id')
            if not name or len(name) > 20 or not pid:
                messages.error(request, 'Name and Pipeline are required.')
                return redirect('lead_stages_index')
            ls.name = name
            ls.pipeline_id = pid
            ls.save()
            messages.success(request, 'Lead Stage successfully updated!')
            return redirect('lead_stages_index')
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=REF, logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=REF, logger=logger)

    @classmethod
    def destroy(cls, request: HttpRequest, lead_stage_id: int) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('delete lead stage'):
                raise PermissionDenied('User lacks permission: delete lead stage')
            ls = get_object_or_404(LeadStage, pk=lead_stage_id)
            ls.delete()
            messages.success(request, 'Lead Stage successfully deleted!')
            return redirect('lead_stages_index')
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=REF, logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=REF, logger=logger)

    @classmethod
    def order(cls, request: HttpRequest) -> JsonResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('manage lead stage'):
                raise PermissionDenied('User lacks permission: manage lead stage')
            order_list = request.POST.getlist('order')
            for idx, item in enumerate(order_list):
                ls = LeadStage.objects.filter(id=item).first()
                if ls:
                    ls.order = idx
                    ls.save()
            return JsonResponse({'success': True})
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=REF, logger=logger, json={})
        except Exception as e:
            return default_undefined_exception(
                request,
                err=e,
                ref=REF,
                logger=logger,
                json='Operation failed.',
                status=500
            )
